modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java

```diff
@@ -79,22 +91,30 @@ public abstract class AbstractKeyValueStorage implements KeyValueStorage {
     protected final AtomicBoolean stopCompaction = new AtomicBoolean();
 
     /** Tracks only cursors, since reading a single entry or a batch is done entirely under {@link #rwLock}. */
-    protected final ReadOperationForCompactionTracker readOperationForCompactionTracker = new ReadOperationForCompactionTracker();
+    protected final ReadOperationForCompactionTracker readOperationForCompactionTracker;
 
-    /**
-     * Used to generate read operation ID for {@link #readOperationForCompactionTracker}.
-     *
-     * <p>Multi-threaded access is guarded by {@link #rwLock}.</p>
-     */
-    protected long readOperationIdGeneratorForTracker;
+    protected final ExecutorService compactionExecutor;
+
+    private final List<CompactionListener> compactionListeners = new CopyOnWriteArrayList<>();
 
     /**
      * Constructor.
      *
      * @param nodeName Node name.
      * @param failureManager Failure processor that is used to handle critical errors.
+     * @param readOperationForCompactionTracker Read operation tracker for metastorage compaction.
+     * @param compactionExecutor Metastorage compaction executor.
      */
-    protected AbstractKeyValueStorage(String nodeName, FailureManager failureManager) {
+    protected AbstractKeyValueStorage(
+            String nodeName,
+            FailureManager failureManager,
+            ReadOperationForCompactionTracker readOperationForCompactionTracker,
+            ExecutorService compactionExecutor
+    ) {
+        this.failureManager = failureManager;
+        this.readOperationForCompactionTracker = readOperationForCompactionTracker;
+        this.compactionExecutor = compactionExecutor;
```

Comment:
Where do you stop this executor?

Reply:
Each implementation must understand what to do with it, for example, for `RocksDbKeyValueStorage` we create the pool ourselves, so we stop it when closing.